DirectXTex/DirectXTexSwizzle.cpp

```diff
@@ -0,0 +1,303 @@
+    const Image* uncompressedDestination;
+    ScratchImage wide;
+    ScratchImage dest;
+    if (IsCompressed(metadata.format))
+    {
+        wide = ScratchImage();
+        HRESULT de = Decompress(srcImages, nimages, metadata, metadata.format, wide);
+        if (FAILED(de))
+            return de;
+        uncompressedSource = wide.GetImages();
+
+        dest = ScratchImage();
+        HRESULT dcom = dest.Initialize2D(uncompressedSource[0].format, srcImages[0].width, srcImages[0].height, nimages, 1);
+        if (FAILED(dcom))
+            return dcom;
+        uncompressedDestination = dest.GetImages();
+    }
+    else
+    {
+        uncompressedSource = srcImages;
+        uncompressedDestination = result.GetImages();
+    }
+
+    for (size_t imageIndex = 0; imageIndex < nimages; imageIndex++)
+    {
+        const uint8_t* sptr = uncompressedSource[imageIndex].pixels;
+        if (!sptr)
+            return E_POINTER;
+
+        uint8_t* dptr = uncompressedDestination[imageIndex].pixels;
+        if (!dptr)
+            return E_POINTER;
+
+        size_t bytesPerPixel = BitsPerPixel(uncompressedSource[imageIndex].format) / 8;
+
+        uint32_t xBytesMask = 0b1010101010101010;
+        uint32_t yBytesMask = 0b0101010101010101;
+
+        for (size_t y = 0; y < uncompressedSource[imageIndex].height; y++)
+        {
+            for (size_t x = 0; x < uncompressedSource[imageIndex].width; x++)
+            {
+                uint32_t swizzleIndex = deposit_bits(x, xBytesMask) + deposit_bits(y, yBytesMask);
+                size_t swizzleOffset = swizzleIndex * bytesPerPixel;
+
+                size_t rowMajorOffset = y * uncompressedSource[0].rowPitch + x * bytesPerPixel;
+
+                size_t sourceOffset = toSwizzle ? rowMajorOffset : swizzleOffset;
+                size_t destOffset = toSwizzle ? swizzleOffset : rowMajorOffset;
+
+                const uint8_t* sourcePixelPointer = sptr + sourceOffset;
+                uint8_t* destPixelPointer = dptr + destOffset;
+                memcpy(destPixelPointer, sourcePixelPointer, bytesPerPixel);
+            }
+        }
+    }
+
+    if (IsCompressed(metadata.format))
+    {
+        TEX_COMPRESS_FLAGS flags = TEX_COMPRESS_DEFAULT; // !!TODO!! actual value
+        float threshold = -42; // !!TODO!! actual value
+
+        Compress(uncompressedDestination, nimages, metadata, metadata.format, flags, threshold, result);
+        wide.Release();
+        dest.Release();
+    }
+
+    return S_OK;
+}
+
+_Use_decl_annotations_
+HRESULT DirectX::StandardSwizzle3D(const Image* srcImages, size_t depth, const TexMetadata& metadata, bool toSwizzle, ScratchImage& result) noexcept
+{
+    HRESULT hr = result.Initialize3D(srcImages[0].format, srcImages[0].width, srcImages[0].height, depth, 1);
+
+    if (!srcImages || !depth || !IsValid(metadata.format) || depth > metadata.depth || !result.GetImages())
+        return E_INVALIDARG;
+
+    if (metadata.IsVolumemap()
```

Comment:
For ``StandardSwizzle3D``, you want:

```
if (!metadata.IsVolumemap() || IsTypeless(metadata.format) || IsPlanar(metadata.format) || IsPalettized(metadata.format)))
    return E_INVALIDARG;
```

and for both you need to remove the ``IsCompressed`` test from this early out since you are adding compress handlnig.

Reply:
duplicate of https://github.com/microsoft/DirectXTex/pull/515#discussion_r1765398888